[Feat-#124] Kakao API를 통한 주소 변환 서비스 계층 구현

zbqmgldjfh opened the pull request:
## Kakao API를 사용한 주소변환
1. 위도 경도를 기반으로 한 주소변환
2. 지역검색을 기반으로 한 주소변환
3. WireMock을 통한 테스트코드 작성

노션에 작성해둔 글
https://lake-peach-ffa.notion.site/Feign-Client-by-Shine-32c0531c2cef4515b10a8340898d1f69

#124

zbqmgldjfh committed:
test: Kakao WireMock을 통한 주소 변환 테스트

zbqmgldjfh committed:
feat: Kakao API를 통한 주소 변환 요청 구현

zbqmgldjfh committed:
test: WireMock을 통한 위경도를 주소로 변환하는 테스트

zbqmgldjfh committed:
feat: 위도, 경도를 주소로 변환 구현 완료

zbqmgldjfh committed:
chore: KakaoAddressResponse 패키지 이동

zbqmgldjfh committed:
refactor(KakaoPlaceApiService#searchAddress): Dto로 받던 검색어를 단순 string으로 받도록 변경

단순 문자열을 DTO로 한번 감싸서 전달하는 불필요한 과정은 필요 없다 생각하였다.

zbqmgldjfh commented:
![image](https://user-images.githubusercontent.com/60593969/208427329-fba9b5a0-e15a-4cfa-89f9-330875c8df44.png)

커버리지가 62.8%나 되었습니다!!

zbqmgldjfh committed:
chore: 모든 테스트 메서드에서 public접근자 제거

pbg0205 commented on src/main/java/com/soolsul/soolsulserver/bar/businees/client/KakaoPlaceApiService.java:
```diff
@@ -0,0 +1,61 @@
+@RequiredArgsConstructor
+public class KakaoPlaceApiService {
+
+    private final KakaoAddressSearchClient addressSearchClient;
```
AddreessSearchClient 와 같이 별도의 컴포넌트를 추가적으로 적용해보는건 어떠세요?  외부 요청하는  RestTemplate 또는 FeignClient 같이 특정 기술에 관련된 컴포넌트의 경우에는 별도의 컴포넌트로 정의해서 의존성을 줄여 변경에 유연한 코드가 될 것 같습니다

zbqmgldjfh replied:
음 도메인을 별도로 분리하자는 의미 이신가요? 아니면 인터페이스로 분리하자는 의미 이신가요???

pbg0205 replied:
PlaceApiService 인터페이스 하위로 KakaoPlaceApiService 를 선언하고  AddressSearchClient 에 상위 인터페이스를 정의하자는 의미였어요!

zbqmgldjfh replied:
넵!! 인터페이스를 분리하여 특정 벤더에 의존하지 않도록 변경하겠습니다 ㅎㅎ

pbg0205 commented on src/test/java/com/soolsul/soolsulserver/bar/businees/client/KakaoPlaceApiServiceTest.java:
```diff
@@ -0,0 +1,52 @@
+
+    @DisplayName("주소를 검색을 통해 장소 목록을 조회한다.")
+    @Test
+    public void lookup_address_by_query() {
```
public 접근 제한자 제거 부탁드려요~

zbqmgldjfh replied:
이부분은 전부 제거 하였습니다 ㅎㅎ, 아마 뒤늦게 push한 커밋에 반영되어 있는것 같습니다 ㅎㅎ

pbg0205 commented on src/test/java/com/soolsul/soolsulserver/bar/businees/client/KakaoPlaceApiServiceTest.java:
```diff
@@ -0,0 +1,52 @@
+                .allSatisfy(address ->
+                        assertThat(address).satisfiesAnyOf(addr -> assertThat(addr.addressName()).contains("전북"))
+                );
+    }
```
`allSatisfy()` 메서드는 처음보네요. 덕분에 하나 알아갑니다👍

zbqmgldjfh replied:
저도 이번에 처음 사용해보게 되었습니다 ㅋㅋㅋ

pbg0205 commented on src/main/java/com/soolsul/soolsulserver/SoolsulServerApplication.java:
```diff
@@ -2,7 +2,9 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.cloud.openfeign.EnableFeignClients;
 
+@EnableFeignClients
```
이전에 @EnableJpaAuditing 과 통일성있게 설정과 관련된 어노테이션은 별도의 @Configuration 으로 분리하는게 좋을 것 같습니다

zbqmgldjfh replied:
넵! 별도의 FeignClientConfiguration 을 분리하고, basepackage를 지정해주었습니다!

pbg0205 commented on src/test/java/com/soolsul/soolsulserver/common/ClientMocksExtention.java:
```diff
@@ -0,0 +1,19 @@
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class ClientMocksExtention implements BeforeAllCallback, AfterAllCallback {
```
Stub 만드는 방식은 한번도 안해봤는데 덕분에 하나 알아갑니다 👍

pbg0205 commented on src/main/java/com/soolsul/soolsulserver/bar/businees/client/dto/KakaoAddressResponse.java:
```diff
@@ -0,0 +1,32 @@
+
+    private String addressName;
+
+    @JsonProperty("region_1depth_name")
```
`@JsonNaming` annotation 을 선언하고 별도의 property value 를 선언하신 이유가 있을까요?

zbqmgldjfh replied:
region1DepthName 의 경우 숫자 1이 중간에 있어서 snake case로 변경시 region_1_depth_name이 되어버리릴까봐 명시적으로 추가해주엇습니다. (이부분은 뇌피셜이며... 그냥 명확하게 에노테이션 달아주자 라고 생각하였습니다 ㅎㅎ)

카카오에서는 "region_1depth_name"으로 값이 넘어오는데, 이를 명확하게 받기 위해 작성해두었습니다 ㅎㅎ

pbg0205 commented on src/main/java/com/soolsul/soolsulserver/common/config/client/KakaoClientAuthHeaderConfiguration.java:
```diff
@@ -0,0 +1,16 @@
+    public RequestInterceptor requestInterceptor() {
+        return requestTemplate -> requestTemplate.header("Authorization", apiToken);
+    }
+}
```
각각이 client의 세부 설정을 위한 @Configuration 을 설정하는건 어떠세요??
Retryer 을 선언 요청 대기시간, 추가 요청 횟수를 설정할 수 있어서 각 client 의 세부 설정할 수 있어서 저희의 정책에 맞게 커스터마이징 할 수 있을 것 같습니다.
아래 학습해본 내용에 관련해서 공유드려요~
- https://github.com/pbg0205/BE-tutorials/blob/main/spring-feign-client/README.md#3-custom-beans-configuration

zbqmgldjfh replied:
이부분은 현재 세부 설정 configuration으로 사용하고 있는것이 맞습니다!!
사용하는 코드쪽에서 보면 다음과 같이 configuration으로 지정하여 사용하고 있습니다.
```
@FeignClient(name = "local-address-client", url = "${api.kakao.api-url}",
        configuration = {KakaoClientAuthHeaderConfiguration.class})
public interface KakaoAddressSearchClient {
    ... 생략...
}
```

이걸 @ configuration을 추가해서 전역으로 설정할 필요는 없는것 같습니다?

말씀하신 것 처럼 retryer, 추가 요청 횟수등같은 경우 전역으로 적용되도록 할필요가 있지만,
위의 Authorization의 경우 모든 응답에 필요한 요소가 아니라 생각됩니다 ㅎㅎ
필요한 client에만 선택적으로 configuration옵션으로 추가하는 쪽이 더 좋은것 같다 생각됩니다 ㅎㅎ

pbg0205 replied:
그러면 KakoClientAuthorizationConfiguration과 @EnableFeignClient 선언 위치가 다른거죠?

zbqmgldjfh replied:
넵 맞습니다!!
<img width="322" alt="image" src="https://user-images.githubusercontent.com/60593969/208801104-2633a8a0-3905-489a-a977-fdccab86126a.png">

pbg0205 replied:
그러면 이대로 유지하는게 좋을거 같아요! 감사합니다🙏

pbg0205 commented on src/test/java/com/soolsul/soolsulserver/common/ClientMocksExtention.java:
```diff
@@ -0,0 +1,19 @@
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class ClientMocksExtention implements BeforeAllCallback, AfterAllCallback {
```
Extension에 s가 빠져있어요~ 추가 부탁드려요!

zbqmgldjfh replied:
변경하였습니다!! 감사합니다!!

pbg0205 commented on src/test/java/com/soolsul/soolsulserver/common/mock/KakaoPlaceMocks.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    private static void setupConvertAddressStub() {
+        addressConvertStub = stubFor(get(urlPathEqualTo(ADDRESS_CONVERT_URL))
```
이 부분 공부해보면서 해당 컨트롤러에 @RequestHeader 추가하고 stubbing 에서 값 없어도 동작하더라구요. 제가 잘못 확인한걸 수도 있는데 혹시나 알게 되면 공유드리겠습니다

pbg0205 commented on src/test/java/com/soolsul/soolsulserver/common/WireMockClientTest.java:
```diff
@@ -0,0 +1,23 @@
+@ExtendWith(ClientMocksExtention.class)
+public @interface WireMockClientTest {
+}
+
```
이 부분 샤인이 적용해보면서 공부해봤는데 덕분에 많이 알아갔습니다 감사합니다

zbqmgldjfh replied:
저도 wiremock에 대하여 처음 알게 되었는데, 서드파티 테스트 하는 용도로 아주 좋은것 같습니다 ㅎㅎ
저도 더욱 애용하게 될것 같아요 ㅎㅎ

pbg0205 replied:
외부 API에 관해서 테스트 격리를 할 수 있어서 좋은 것 같아요😃 생각해보니까 외부 API에 관한 예외처리도 추가하면 좋을 것 같아요. 특정 예외 던져준다고 어렴풋이 봤는데 저도 한번 찾아볼게요. 이 부분은 별도 일감으로 등록해둘게요~

pbg0205 commented on src/test/java/com/soolsul/soolsulserver/common/mock/KakaoPlaceMocks.java:
```diff
@@ -0,0 +1,55 @@
+// https://www.baeldung.com/spring-cloud-feign-integration-tests
+public abstract class KakaoPlaceMocks {
+
+    private final static String ADDRESS_SEARCH_URL = "/v2/local/search/address.json";
```
static final 순서 변경 부탁드려요~

zbqmgldjfh replied:
앗 ㅋㅋㅋㅋ 수정하였습니다 ㅎㅎㅎ

pbg0205 reviewed (approved):
샤인 수고 많으셨습니다~ OpenFeign 는 몰랐던 내용인데 덕분에 이번에 공부해보게 되었네요 감사합니다. 인사이트가 대단하시네요👍 아래 리뷰하기 위해 잠깐 공부했던 내용 관련해서 공유드립니다.
리뷰에 개인적인 의견이나 코멘트 남겼습니다 확인부탁드려요~
- https://github.com/pbg0205/BE-tutorials/blob/main/spring-feign-client/README.md#3-custom-beans-configuration

zbqmgldjfh committed:
refactor: FeignClientConfiguration 분리

zbqmgldjfh committed:
chore: 오타 수정

zbqmgldjfh committed:
feat: Retryer 추가 구현

0.5초의 간격으로 시작해 최대 3초의 간격으로 점점 증가하며, 최대5번 재시도한다.

zbqmgldjfh committed:
feat: PlaceApiService, AddressSearchClient 인터페이스로 분리

zbqmgldjfh committed:
test: 인수테스트에서 메시지 내용 검증 제거하기

과도한 검증을 지양하며, 메시지는 빈번하게 변경될 여지가 있기 때문에 해당 부분의 검증을 전부 제거하며, 커스텀 상태코드 검증으로 충분하다 판단됨
